File: packages/server/simple/rust/src/cronjobs.rs
use std::str::FromStr;
use std::sync::Arc;

use chrono::Utc;
use cron::Schedule;
use serde_json::Value;
use uuid::Uuid;

use crate::db;
use crate::handlers::AppState;
use crate::models::*;

pub const EVENT_CRONJOB_CREATE: &str = "cronjob.create";
pub const EVENT_CRONJOB_UPDATE: &str = "cronjob.update";
pub const EVENT_CRONJOB_DELETE: &str = "cronjob.delete";

pub fn webhook_cronjob_data(job: &CronJob) -> Value {
    serde_json::json!({"cronjob": job})
}

pub async fn start_cron_scheduler(state: Arc<AppState>) {
    tokio::spawn(async move {
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(30));
        loop {
            interval.tick().await;
            let jobs = {
                let conn = match state.db.lock() {
                    Ok(c) => c,
                    Err(_) => continue,
                };
                db::list_cron_jobs(&conn).unwrap_or_default()
            };
            for job in jobs {
                if !job.is_active {
                    continue;
                }
                if job.schedule.is_empty() {
                    continue;
                }
                let schedule = match Schedule::from_str(&job.schedule) {
                    Ok(s) => s,
                    Err(_) => continue,
                };
                let now = Utc::now();
                let last_run = if job.last_run_at.is_empty() {
                    now - chrono::Duration::hours(1)
                } else {
                    chrono::DateTime::parse_from_rfc3339(&job.last_run_at)
                        .map(|d| d.with_timezone(&chrono::Utc))
                        .unwrap_or(now - chrono::Duration::hours(1))
                };
                if let Some(next_run) = schedule.after(&last_run).next() {
                    if next_run <= now {
                        execute_cron_job(Arc::clone(&state), job).await;
                    }
                }
            }
        }
    });
}

pub async fn execute_cron_job(state: Arc<AppState>, job: CronJob) {
    let started_at = Utc::now().to_rfc3339();
    let client = &state.http_client;

    let method = job.method.to_uppercase();
    let mut req_builder = match method.as_str() {
        "GET" => client.get(&job.command),
        "POST" => client.post(&job.command),
        "PUT" => client.put(&job.command),
        "PATCH" => client.patch(&job.command),
        "DELETE" => client.delete(&job.command),
        _ => client.get(&job.command),
    };

    if !job.headers.is_empty() {
        if let Ok(headers) = serde_json::from_str::<Vec<Vec<String>>>(&job.headers) {
            for h in headers {
                if h.len() == 2 {
                    req_builder = req_builder.header(&h[0], &h[1]);
                }
            }
        }
    }

    if !job.body.is_empty() && (method == "POST" || method == "PUT" || method == "PATCH") {
        req_builder = req_builder.body(job.body.clone());
    }

    let (status, output, error) = match req_builder.send().await {
        Ok(resp) => {
            let sc = resp.status().as_u16();
            let body = resp.text().await.unwrap_or_default();
            (sc as i64, body, String::new())
        }
        Err(e) => (0, String::new(), e.to_string()),
    };

    let finished_at = Utc::now().to_rfc3339();
    let started = chrono::DateTime::parse_from_rfc3339(&started_at)
        .map(|d| d.with_timezone(&chrono::Utc))
        .unwrap_or(Utc::now());
    let finished = chrono::DateTime::parse_from_rfc3339(&finished_at)
        .map(|d| d.with_timezone(&chrono::Utc))
        .unwrap_or(Utc::now());
    let duration_ms = (finished - started).num_milliseconds();

    let status_str = if (200..300).contains(&status) { "success" } else { "failure" };

    let log = CronJobLog {
        id: Uuid::new_v4().to_string().replace('-', ""),
        cronjob_id: job.id.clone(),
        started_at,
        finished_at,
        duration_ms,
        status: status_str.to_string(),
        output,
        error,
    };

    {
        let conn = match state.db.lock() {
            Ok(c) => c,
            Err(_) => return,
        };
        let _ = db::insert_cron_job_log(&conn, &log);
        let _ = db::update_cron_job_last_run(&conn, &job.id, &log.finished_at, status_str);
    }
}

File: packages/server/backbone/go/src/cronjobs.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/robfig/cron/v3"
)

const (
	EventCronjobCreate = "cronjob.create"
	EventCronjobUpdate = "cronjob.update"
	EventCronjobDelete = "cronjob.delete"
)

type CronJob struct {
	ID            string `json:"id"`
	Name          string `json:"name"`
	Schedule      string `json:"schedule"`
	Command       string `json:"command"`
	Method        string `json:"method"`
	Headers       string `json:"headers"`
	Body          string `json:"body"`
	IsActive      bool   `json:"is_active"`
	LastRunAt     string `json:"last_run_at"`
	LastRunStatus string `json:"last_run_status"`
	CreatedAt     string `json:"created_at"`
	UpdatedAt     string `json:"updated_at"`
}

type CronJobLog struct {
	ID         string `json:"id"`
	CronJobID  string `json:"cronjob_id"`
	StartedAt  string `json:"started_at"`
	FinishedAt string `json:"finished_at"`
	DurationMs int64  `json:"duration_ms"`
	Status     string `json:"status"`
	Output     string `json:"output"`
	Error      string `json:"error"`
}

func startCronScheduler(db *sql.DB) *cron.Cron {
	c := cron.New(cron.WithSeconds())
	c.AddFunc("@every 30s", func() {
		jobs, err := listCronJobs(db)
		if err != nil {
			log.Printf("cron scheduler: list jobs: %v", err)
			return
		}
		now := time.Now()
		for _, j := range jobs {
			if !j.IsActive {
				continue
			}
			entry, err := cron.ParseStandard(j.Schedule)
			if err != nil {
				continue
			}
			lastRun, _ := time.Parse(time.RFC3339, j.LastRunAt)
			nextRun := entry.Next(lastRun)
			if nextRun.IsZero() {
				continue
			}
			if !now.Before(nextRun) {
				go executeCronJob(db, j)
			}
		}
	})
	c.Start()
	return c
}

func executeCronJob(db *sql.DB, job CronJob) {
	startedAt := time.Now().UTC().Format(time.RFC3339)
	var status, output, errMsg string
	var durationMs int64

	method := strings.ToUpper(job.Method)
	if method == "" {
		method = "GET"
	}

	var bodyReader io.Reader
	if job.Body != "" {
		bodyReader = strings.NewReader(job.Body)
	}

	req, err := http.NewRequest(method, job.Command, bodyReader)
	if err != nil {
		errMsg = fmt.Sprintf("create request: %v", err)
		status = "failure"
	} else {
		if job.Headers != "" {
			var headers map[string]string
			if err := json.Unmarshal([]byte(job.Headers), &headers); err == nil {
				for k, v := range headers {
					req.Header.Set(k, v)
				}
			}
		}
		client := &http.Client{Timeout: 30 * time.Second}
		resp, err := client.Do(req)
		finishedAt := time.Now()
		if startedTime, parseErr := time.Parse(time.RFC3339, startedAt); parseErr == nil {
			durationMs = finishedAt.Sub(startedTime).Milliseconds()
		}

		if err != nil {
			errMsg = err.Error()
			status = "failure"
		} else {
			defer resp.Body.Close()
			body, _ := io.ReadAll(resp.Body)
			output = string(body)
			if len(output) > 10240 {
				output = output[:10240]
			}
			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
				status = "success"
			} else {
				status = "failure"
				errMsg = fmt.Sprintf("HTTP %d", resp.StatusCode)
			}
		}
	}

	now := time.Now().UTC()
	id := generateID()
	logEntry := &CronJobLog{
		ID:         id,
		CronJobID:  job.ID,
		StartedAt:  startedAt,
		FinishedAt: now.Format(time.RFC3339),
		DurationMs: durationMs,
		Status:     status,
		Output:     output,
		Error:      errMsg,
	}
	updateCronJobLastRun(db, job.ID, now.Format(time.RFC3339), status)
	insertCronJobLog(db, logEntry)
}

func webhookCronjobData(job *CronJob) map[string]any {
	return map[string]any{
		"cronjob": job,
	}
}
